File: Reqnroll/Chapter16/16.3-EnsurePattern/After/WIMP.Specs/StepDefinitions/CustomerCollectionStepDefinitions.cs
using Reqnroll;

using WIMP.App.Models;
using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class CustomerCollectionStepDefinitions(
    AuthenticationApiDriver authApiDriver,
    OrderingApiDriver orderingApiDriver,
    AuthenticationContext authenticationContext,
    OrderingContext orderingContext)
{
    private OrderCollectionDetails? orderCollectionDetails;

    [Given("the customer is authenticated")]
    public async Task GivenTheCustomerIsAuthenticated()
    {
        await authApiDriver
            .Login(DomainDefaults.CustomerName, DomainDefaults.Password)
            .Execute();
        authenticationContext.LoggedInCustomerName = DomainDefaults.CustomerName;
    }

    [Given("they have placed an order for {int} pizzas")]
    public async Task GivenTheyHavePlacedAnOrderForPizzas(int count)
    {
        var orderRequest = new PlaceOrderRequestObjectMother()
            .WithItems(count)
            .Build();
        var placedOrder = await orderingApiDriver
            .PlaceOrder(orderRequest)
            .Execute();
        orderingContext.PlacedOrderNo = placedOrder.OrderNo;
    }

    [When("they choose to collect their order")]
    public async Task WhenTheyChooseToCollectTheirOrder()
    {
        await orderingContext.EnsureOrderPlaced();

        orderCollectionDetails = await orderingApiDriver
            .SetForCollection(orderingContext.PlacedOrderNo ?? throw new InvalidOperationException("No order placed."))
            .Execute();
    }

    [Then("they should be asked to confirm contact details")]
    public void ThenTheyShouldBeAskedToConfirmContactDetails()
    {
        Assert.IsNotNull(orderCollectionDetails, "The order was not set to customer-collection");
        Assert.IsTrue(orderCollectionDetails.ContactDetailsConfirmationRequested);
    }

    [Then("a collection receipt should be printed with")]
    public void ThenACollectionReceiptShouldBePrintedWith(DataTable dataTable)
    {
        Assert.IsNotNull(orderCollectionDetails, "The order was not set to customer-collection");
        int expectedBoxes = int.Parse(dataTable.Rows[0]["boxes to be collected"]);
        Assert.AreEqual(expectedBoxes, orderCollectionDetails.BoxesToBeCollected, "Invalid number of boxes for customer-collection");
    }
}

File: CucumberJVM/Chapter16/16.3-EnsurePattern/After/src/test/java/com/wimp/app/specs/stepdefinitions/CustomerCollectionStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.models.OrderCollectionDetails;
import com.wimp.app.specs.drivers.OrderingApiDriver;
import com.wimp.app.specs.support.DataTableDiffHelper;
import com.wimp.app.specs.support.OrderingContext;
import io.cucumber.datatable.DataTable;
import io.cucumber.java.en.*;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

public class CustomerCollectionStepDefinitions {
    private final OrderingContext orderingContext;
    private final OrderingApiDriver orderingApiDriver;
    private OrderCollectionDetails orderCollectionDetails;

    public CustomerCollectionStepDefinitions(OrderingContext orderingContext, OrderingApiDriver orderingApiDriver) {
        this.orderingContext = orderingContext;
        this.orderingApiDriver = orderingApiDriver;
    }

    @When("they choose to collect their order")
    public void theyChooseToCollectTheirOrder() throws Exception {
        orderingContext.ensureOrderPlaced();

        orderCollectionDetails = orderingApiDriver.setForCollection(orderingContext.getCurrentOrderNoVerified()).execute();
    }

    @Then("they should be asked to confirm contact details")
    public void theyShouldBeAskedToConfirmContactDetails() {
        assertNotNull(orderCollectionDetails, "The order was not set to customer-collection");
        assertTrue(orderCollectionDetails.isContactDetailsConfirmationRequested());
    }

    @Then("a collection receipt should be printed with")
    public void aCollectionReceiptShouldBePrintedWith(DataTable expectedCollectionDetailsDataTable) {
        assertNotNull(orderCollectionDetails, "The order was not set to customer-collection");
        var actualCollectionDetailsTable = DataTableDiffHelper.createDataTableWithHeader(List.of(orderCollectionDetails), expectedCollectionDetailsDataTable);
        expectedCollectionDetailsDataTable.diff(actualCollectionDetailsTable);
    }
}
